test/index.test.js

```diff
@@ -241,6 +241,19 @@ describe('config parser', () => {
                             /YAMLException: Cannot have nonexistent job in stages: publish,deploy/
                         );
                     }));
+
+                it('returns an error if a job within a stage is triggered by jobs that are not within the same stage', () =>
```

Comment:
Can you also add a test for below scenario?

1. stage job does not have `requires`, but uses a job template which has `requires` having invalid trigger (Ex: `~commit`) for stage job.

Reply:
So I added a test of job using a template that has `[~pr, ~commit]` as trigger, test is passing because apparently, the trigger is replaced with the setup job. So in the case of job `boom` that extends this template, the final config is:
```
  boom: {
    image: 'node:4',
    steps: [ [Object], [Object] ],
    environment: {
      FOO: 'from template',
      BAR: 'foo',
      SD_TEMPLATE_FULLNAME: 'mytemplate',
      SD_TEMPLATE_NAME: 'mytemplate',
      SD_TEMPLATE_NAMESPACE: '',
      SD_TEMPLATE_VERSION: '1.2.3'
    },
    secrets: [ 'GIT_KEY' ],
    settings: { email: 'foo@example.com' },
    parameters: undefined,
    annotations: {},
    requires: [ '~stage@canary:setup' ],
    stage: { name: 'canary' },
    sourcePaths: [],
    templateId: 7754,
    commands: [ [Object], [Object] ]
  }
```
Is this by design or is this a bug?

Reply:
After checking, I think this is designed as such that the [stages is flattened](https://github.com/screwdriver-cd/config-parser/blob/5a0c19b985eef25d6e53cf1f2e308e44ce61486b/lib/phase/flatten.js#L686-L692) before the template flattened. So during the stages flattened, since there is no requires in `boom` job, it is immediately replaced it with setup job.